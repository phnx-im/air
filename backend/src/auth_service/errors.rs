// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::openmls::prelude::KeyPackageVerifyError;
use thiserror::Error;

/// Error fetching a message from the QS.
#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum AsDequeueError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Couldn't find the requested queue.
    #[error("Couldn't find the requested queue")]
    QueueNotFound,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum InitUserRegistrationError {
    /// Library error
    #[error("Library error")]
    LibraryError,
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// User already exists
    #[error("User already exists")]
    UserAlreadyExists,
    /// Invalid CSR
    #[error("Invalid CSR")]
    InvalidCsr,
    /// Error during OPAQUE registration
    #[error("Error during OPAQUE registration")]
    OpaqueRegistrationFailed,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum FinishUserRegistrationError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Client credential not found
    #[error("Client credential not found")]
    ClientCredentialNotFound,
    /// Error finishing OPAQUE login handshake
    #[error("Error finishing OPAQUE login handshake")]
    OpaqueLoginFinishFailed,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum DeleteUserError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum UserClientsError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum InitClientAdditionError {
    /// Library error
    #[error("Library error")]
    LibraryError,
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Client already exists
    #[error("Client already exists")]
    ClientAlreadyExists,
    /// Invalid CSR
    #[error("Invalid CSR")]
    InvalidCsr,
    /// Error during OPAQUE login handshake
    #[error("Error during OPAQUE login handshake")]
    OpaqueLoginFailed,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum FinishClientAdditionError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Client credential not found
    #[error("Client credential not found")]
    ClientCredentialNotFound,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum DeleteClientError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, Clone)]
pub enum PublishKeyPackageError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Invalid KeyPackage
    #[error(transparent)]
    KeyPackageValidationError(#[from] KeyPackageVerifyError),
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum ClientKeyPackageError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum UserKeyPackagesError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum EnqueueMessageError {
    /// Library error
    #[error("Library error")]
    LibraryError,
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Client not found
    #[error("Client not found")]
    ClientNotFound,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum IssueTokensError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Too many tokens
    #[error("Too many tokens")]
    TooManyTokens,
    /// PrivacyPass protocol error
    #[error("PrivacyPass protocol error")]
    PrivacyPassError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum Init2FactorAuthError {
    /// Library error
    #[error("Library error")]
    LibraryError,
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Error during OPAQUE login handshake
    #[error("Error during OPAQUE login handshake")]
    OpaqueLoginFailed,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum AsCredentialsError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
}

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum AsVerificationError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Could not find client
    #[error("Could not find client")]
    UnknownClient,
    /// Could not find user
    #[error("Could not find user")]
    UnknownUser,
    /// Could not authenticate message
    #[error("Could not authenticate message")]
    AuthenticationFailed,
}

#[derive(Error, Debug, Clone)]
pub enum AsProcessingError {
    /// Authentication error
    #[error(transparent)]
    AuthenticationError(#[from] AsVerificationError),
    #[error(transparent)]
    AsDequeueError(#[from] AsDequeueError),
    #[error(transparent)]
    InitUserRegistrationError(#[from] InitUserRegistrationError),
    #[error(transparent)]
    FinishUserRegistrationError(#[from] FinishUserRegistrationError),
    #[error(transparent)]
    DeleteUserError(#[from] DeleteUserError),
    #[error(transparent)]
    UserClientsError(#[from] UserClientsError),
    #[error(transparent)]
    InitClientAdditionError(#[from] InitClientAdditionError),
    #[error(transparent)]
    FinishClientAdditionError(#[from] FinishClientAdditionError),
    #[error(transparent)]
    DeleteClientError(#[from] DeleteClientError),
    #[error(transparent)]
    PublishKeyPackageError(#[from] PublishKeyPackageError),
    #[error(transparent)]
    ClientKeyPackageError(#[from] ClientKeyPackageError),
    #[error(transparent)]
    UserKeyPackagesError(#[from] UserKeyPackagesError),
    #[error(transparent)]
    EnqueueMessageError(#[from] EnqueueMessageError),
    #[error(transparent)]
    IssueTokensError(#[from] IssueTokensError),
    #[error(transparent)]
    Init2FactorAuthError(#[from] Init2FactorAuthError),
    #[error(transparent)]
    AsCredentialsError(#[from] AsCredentialsError),
}
