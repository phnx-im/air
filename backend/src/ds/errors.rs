use mls_assist::GroupEpoch;
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Error, Serialize, Deserialize)]
pub enum GroupCreationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid group id
    #[error("Invalid group id")]
    InvalidGroupId, // Wrong group ID [too explicit]
    // TODO: Probably a bit too vague. Not making this transparent for now, as we'd have to derive serde for a number of OpenMLS errors.
    /// Invalid group creation parameters.
    #[error("Invalid group creation parameters.")]
    InvalidParameters,
    /// Error storing group state.
    #[error("Error storing group state.")]
    StorageError,
    /// Could not encrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotEncrypt,
}

/// Error struct for the verification of [`UnverifiedGroupOperationParams`]
/// instances.
#[derive(Debug, Error, Serialize, Deserialize)]
pub enum GroupOperationVerificationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Wrong group ID
    #[error("Wrong group ID")]
    WrongGroupId, // Wrong group ID [too explicit]
    /// Group expired
    #[error("Group expired")]
    GroupExpired, // Group expired [too explicit]
    /// Failed to decrypt roster with given key
    #[error("Failed to decrypt roster with given key")]
    WrongRosterKey,
    /// No client with the given index could be found in roster
    #[error("No client with the given index could be found in roster")]
    UnknownSender,
    /// The sender has not yet committed to the group and thus can't send application messages
    #[error(
        "The sender has not yet committed to the group and thus can't send application messages"
    )]
    UninitializedSender,
    /// Failed to authenticate sender
    #[error("Failed to authenticate sender")]
    MemberAuthenticationFailed,
}

/// Error struct for processing of DS group operations.
#[derive(Debug, Error, Serialize, Deserialize)]
pub enum GroupOperationProcessingError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Request verification failed
    #[error("Request verification failed")]
    RequestVerificationFailure(GroupOperationVerificationError),
    /// Failed to send non-commit message
    #[error("Failed to send non-commit message")]
    NonCommitSendingFailure(DistributeNonCommitError),
    /// Failed to send commit message
    #[error("Failed to send commit message")]
    CommitSendingFailure(ProcessCommitError),
    /// Failed to update queue config
    #[error("Failed to update queue config")]
    QueueConfigUpdateFailure(UpdateQueueConfigError),
}

impl From<GroupOperationVerificationError> for GroupOperationProcessingError {
    fn from(e: GroupOperationVerificationError) -> Self {
        if matches!(e, GroupOperationVerificationError::LibraryError) {
            Self::LibraryError
        } else {
            Self::RequestVerificationFailure(e)
        }
    }
}

impl From<DistributeNonCommitError> for GroupOperationProcessingError {
    fn from(e: DistributeNonCommitError) -> Self {
        if matches!(e, DistributeNonCommitError::LibraryError) {
            Self::LibraryError
        } else {
            Self::NonCommitSendingFailure(e)
        }
    }
}

/// Group creation error.
#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum CreateGroupError {
    /// Library Error
    #[error("LibraryError")]
    LibraryError,
    /// The signature is invalid.
    #[error("The signature is invalid.")]
    InvalidSignature,
    /// Insufficient randomness
    #[error("Insufficient randomness.")]
    InsufficientRandomness,
    /// Storage provider failed to store group state.
    #[error("Storage provider failed to store group state.")]
    StorageError,
}

/// Group encryption error.
#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum EncryptGroupError {
    /// Encryption error
    #[error("Encryption error")]
    EncryptionError,
    /// Insufficient randomness to encrypt
    #[error("Insufficient randomness to encrypt")]
    InsufficientRandomness,
}

/// Error updating queue config.
#[derive(Error, Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub enum UpdateQueueConfigError {
    /// LibraryError
    #[error("LibraryError")]
    LibraryError,
    /// Insufficient randomness
    #[error("Insufficient randomness.")]
    InsufficientRandomness,
    /// Error storing updated group state
    #[error("Error storing updated group state.")]
    StorageError,
    /// Couldn't find sender in the roster
    #[error("Couldn't find sender in the roster")]
    UnknownSender,
}

/// Error sending non-commit messages.
#[derive(Error, Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub enum ProcessCommitError {
    /// There is a mismatch between the message epoch and the group epoch.
    #[error("There is a mismatch between the message epoch and the group epoch.")]
    EpochMismatch,
    /// The sender does not have sufficient privileges for the desired roster operation.
    #[error("The sender does not have sufficient privileges for the desired roster operation.")]
    InsufficientPrivileges,
    /// Could not apply the given modifications to the roster.
    #[error("Could not apply the given modifications to the roster.")]
    InvalidRosterDelta,
    /// Error distributing the message.
    #[error("Error distributing the message.")]
    DistributionError,
    /// LibraryError
    #[error("LibraryError")]
    LibraryError,
    /// Insufficient randomness
    #[error("Insufficient randomness.")]
    InsufficientRandomness,
    /// Error storing updated group state
    #[error("Error storing updated group state.")]
    StorageError,
}

/// Error distributing messages.
#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub(super) enum MessageDistributionError {
    /// Error delivering the message to the QS
    #[error("DeliveryError")]
    DeliveryError,
}

/// Potential errors when distributing a non-commit message.
#[derive(Debug, Error, Serialize, Deserialize)]
pub enum DistributeNonCommitError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Failed to distribute message to other members
    #[error("Failed to distribute message to other members")]
    DistributionError,
    /// Message epoch doesn't match group epoch
    #[error("Message epoch doesn't match group epoch")]
    WrongEpoch(GroupEpoch, GroupEpoch),
}

/// Potential errors when adding a user.
#[derive(Debug, Error, Serialize, Deserialize)]
pub enum UserAdditionError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
}

/// Potential errors when processing a message.
#[derive(Debug, Error, Serialize, Deserialize)]
pub enum DsProcessingError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Failed to distribute message to other members
    #[error("Failed to distribute message to other members")]
    DistributionError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Invalid signature.
    #[error("Invalid signature.")]
    InvalidSignature,
    /// Group not found.
    #[error("Group not found.")]
    GroupNotFound,
    /// Could not decrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotDecrypt,
    /// Could not encrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotEncrypt,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    /// Invalid sender type.
    #[error("Invalid sender type")]
    InvalidSenderType,
    /// Error storing encrypted group state.
    #[error("Error storing encrypted group state.")]
    StorageError,
    /// Error adding user.
    #[error(transparent)]
    AddUserError(#[from] UserAdditionError),
}
