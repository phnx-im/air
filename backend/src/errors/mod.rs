// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    group, memory_provider::MlsAssistMemoryStorage, openmls::group::MergeCommitError,
};
use thiserror::Error;
use tonic::Status;
use tracing::error;

use phnxtypes::codec::PhnxCodec;

pub(crate) mod auth_service;
pub(crate) mod qs;

pub(crate) type CborMlsAssistStorage = MlsAssistMemoryStorage<PhnxCodec>;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] phnxtypes::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for Status {
    fn from(error: StorageError) -> Self {
        error!(%error, "storage error");
        match error {
            StorageError::Database(_) => Self::internal("Database error"),
            StorageError::Serde(_) => Self::internal("Seriazation error"),
        }
    }
}
#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug)]
pub(super) enum QueueError {
    #[error(transparent)]
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers.
    #[error("Mismatching sequence numbers.")]
    SequenceNumberMismatch,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<phnxtypes::codec::Error> for QueueError {
    fn from(e: phnxtypes::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package.")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupOperationError> for Status {
    fn from(_e: GroupOperationError) -> Self {
        todo!()
    }
}

/// Potential errors when updating a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum ClientUpdateError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ClientUpdateError> for Status {
    fn from(_e: ClientUpdateError) -> Self {
        todo!()
    }
}

/// Potential errors when joining a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum JoinGroupError {
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum JoinConnectionGroupError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Not a connection group.
    #[error("Not a connection group.")]
    NotAConnectionGroup,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<JoinConnectionGroupError> for Status {
    fn from(_e: JoinConnectionGroupError) -> Self {
        todo!()
    }
}

/// Potential errors when adding a user.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum ClientAdditionError {
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when removing clients.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum ClientRemovalError {
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum GroupDeletionError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupDeletionError> for Status {
    fn from(_e: GroupDeletionError) -> Self {
        todo!()
    }
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum ClientSelfRemovalError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ClientSelfRemovalError> for Status {
    fn from(_e: ClientSelfRemovalError) -> Self {
        todo!()
    }
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub(crate) enum ResyncClientError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ResyncClientError> for Status {
    fn from(_e: ResyncClientError) -> Self {
        todo!()
    }
}
