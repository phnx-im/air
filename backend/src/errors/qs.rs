// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use thiserror::Error;
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

/// Error fetching a message from the QS.
#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsDequeueError {
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Couldn't find the requested queue.
    #[error("Couldn't find the requested queue")]
    QueueNotFound,
}

// === Client ===

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsCreateClientRecordError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Error creating client record
    #[error("Error creating user record")]
    StorageError,
    /// Invalid KeyPackage
    #[error("Invalid KeyPackage")]
    InvalidKeyPackage,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsUpdateClientRecordError {
    /// Client not found
    #[error("Client not found")]
    UnknownClient,
    /// Error creating client record
    #[error("Error creating user record")]
    StorageError,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsGetClientError {
    /// Error getting client record
    #[error("Error getting client record")]
    StorageError,
}

// === User ===

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsCreateUserError {
    /// Error creating client record
    #[error("Error creating user record")]
    StorageError,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsUpdateUserError {
    /// User not found
    #[error("User not found")]
    UnknownUser,
    /// Error updating user record
    #[error("Error updating user record")]
    StorageError,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsGetUserError {
    /// Error getting user record
    #[error("Error getting user record")]
    StorageError,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsDeleteUserError {
    /// Error deleteing user record
    #[error("Error deleteing user record")]
    StorageError,
}

// === Key Packages ===

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsPublishKeyPackagesError {
    #[error("Library Error")]
    LibraryError,
    /// Error publishing key packages
    #[error("Error publishing key packages")]
    StorageError,
    /// Invalid KeyPackage
    #[error("Invalid KeyPackage")]
    InvalidKeyPackage,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsClientKeyPackageError {
    /// Error retrieving client key package
    #[error("Error retrieving client key package")]
    StorageError,
    /// No KeyPackages are available
    #[error("No KeyPackages are available")]
    NoKeyPackages,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsKeyPackageError {
    /// Library error
    #[error("Library Error")]
    LibraryError,
    /// Decryption error
    #[error("Decryption error")]
    DecryptionError,
    /// Invalid KeyPackage
    #[error("Invalid KeyPackage")]
    InvalidKeyPackage,
    /// Error retrieving user key packages
    #[error("Error retrieving user key packages")]
    StorageError,
}

#[derive(Error, Debug, Clone, TlsSerialize, TlsDeserializeBytes, TlsSize)]
#[repr(u8)]
pub enum QsEncryptionKeyError {
    /// Library error
    #[error("Library Error")]
    LibraryError,
    /// Error retrieving encryption key
    #[error("Error retrieving encryption key")]
    StorageError,
}
