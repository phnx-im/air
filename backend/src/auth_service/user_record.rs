// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use opaque_ke::ServerRegistration;
use phnxtypes::{crypto::OpaqueCiphersuite, identifiers::QualifiedUserName};

use crate::errors::StorageError;

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(PartialEq, Eq))]
pub(super) struct UserRecord {
    user_name: QualifiedUserName,
    password_file: ServerRegistration<OpaqueCiphersuite>,
}

impl UserRecord {
    fn new(
        user_name: QualifiedUserName,
        password_file: ServerRegistration<OpaqueCiphersuite>,
    ) -> Self {
        Self {
            user_name,
            password_file,
        }
    }

    pub(super) async fn new_and_store(
        connection: impl sqlx::PgExecutor<'_>,
        user_name: &QualifiedUserName,
        opaque_record: &ServerRegistration<OpaqueCiphersuite>,
    ) -> Result<Self, StorageError> {
        let user_record = Self::new(user_name.clone(), opaque_record.clone());
        user_record.store(connection).await?;
        Ok(user_record)
    }

    #[cfg(test)]
    pub(super) fn user_name(&self) -> &QualifiedUserName {
        &self.user_name
    }

    pub(super) fn into_password_file(self) -> ServerRegistration<OpaqueCiphersuite> {
        self.password_file
    }
}

pub(crate) mod persistence {
    use phnxtypes::{
        codec::PhnxCodec,
        identifiers::{QualifiedUserName, UserName},
    };
    use sqlx::PgExecutor;

    use crate::errors::StorageError;

    use super::UserRecord;

    impl UserRecord {
        /// Loads the AsUserRecord for a given UserName. Returns None if no AsUserRecord
        /// exists for the given UserId.
        pub(in crate::auth_service) async fn load(
            connection: impl PgExecutor<'_>,
            user_name: &QualifiedUserName,
        ) -> Result<Option<UserRecord>, StorageError> {
            sqlx::query!(
                r#"SELECT user_name as "user_name: UserName", password_file
                FROM as_user_records
                WHERE user_name = $1"#,
                user_name.to_string(),
            )
            .fetch_optional(connection)
            .await?
            .map(|record| {
                let password_file = PhnxCodec::from_slice(&record.password_file)?;
                Ok(UserRecord::new(user_name.clone(), password_file))
            })
            .transpose()
        }

        /// Create a new user with the given user name. If a user with the given user
        /// name already exists, an error is returned.
        pub(super) async fn store(
            &self,
            connection: impl PgExecutor<'_>,
        ) -> Result<(), StorageError> {
            let password_file_bytes = PhnxCodec::to_vec(&self.password_file)?;
            sqlx::query!(
                "INSERT INTO as_user_records (user_name, password_file) VALUES ($1, $2)",
                self.user_name.to_string(),
                password_file_bytes,
            )
            .execute(connection)
            .await?;
            Ok(())
        }

        /// Deletes the AsUserRecord for a given UserId. Returns true if a AsUserRecord
        /// was deleted, false if no AsUserRecord existed for the given UserId.
        ///
        /// The storage provider must also delete the following:
        ///  - All clients of the user
        ///  - All enqueued messages for the respective clients
        ///  - All key packages for the respective clients
        pub(in crate::auth_service) async fn delete(
            connection: impl PgExecutor<'_>,
            user_name: &QualifiedUserName,
        ) -> Result<(), sqlx::Error> {
            // The database cascades the delete to the clients and their connection packages.
            sqlx::query!(
                "DELETE FROM as_user_records WHERE user_name = $1",
                user_name.to_string()
            )
            .execute(connection)
            .await?;
            Ok(())
        }
    }

    #[cfg(test)]
    pub(crate) mod tests {
        use opaque_ke::{
            ClientRegistration, ClientRegistrationFinishParameters, ServerRegistration, ServerSetup,
        };
        use sqlx::PgPool;
        use uuid::Uuid;

        use super::*;

        pub(crate) async fn store_random_user_record(pool: &PgPool) -> anyhow::Result<UserRecord> {
            let user_name: QualifiedUserName = format!("{}@example.com", Uuid::new_v4()).parse()?;
            let password = b"password";

            let mut rng = rand::thread_rng();
            let server_setup = ServerSetup::new(&mut rng);
            let client_registration_start_result =
                ClientRegistration::start(&mut rng, password).unwrap();
            let server_registration_start_result = ServerRegistration::start(
                &server_setup,
                client_registration_start_result.message,
                user_name.to_string().as_bytes(),
            )
            .unwrap();
            let client_registration_finish_result = client_registration_start_result
                .state
                .finish(
                    &mut rng,
                    password,
                    server_registration_start_result.message,
                    ClientRegistrationFinishParameters::default(),
                )
                .unwrap();
            let password_file =
                ServerRegistration::finish(client_registration_finish_result.message);

            let record = UserRecord {
                user_name,
                password_file,
            };
            record.store(pool).await?;
            Ok(record)
        }

        #[sqlx::test]
        async fn load(pool: PgPool) -> anyhow::Result<()> {
            let user_record = store_random_user_record(&pool).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name)
                .await?
                .expect("missing user record");
            assert_eq!(loaded, user_record);

            Ok(())
        }

        #[sqlx::test]
        async fn delete(pool: PgPool) -> anyhow::Result<()> {
            let user_record = store_random_user_record(&pool).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name)
                .await?
                .expect("missing user record");
            assert_eq!(loaded, user_record);

            UserRecord::delete(&pool, &user_record.user_name).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name).await?;
            assert!(loaded.is_none());

            Ok(())
        }
    }
}
