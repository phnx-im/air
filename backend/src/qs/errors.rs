// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::errors::StorageError;

use super::network_provider::NetworkProvider;
use phnxtypes::crypto::errors::{DecryptionError, KeyGenerationError};
use thiserror::Error;

// === DS API errors ===

/// Error fetching a message from the QS.
#[derive(Error, Debug)]
pub enum QsEnqueueError<N: NetworkProvider> {
    /// Couldn't find the requested queue.
    #[error("Couldn't find the requested queue")]
    QueueNotFound,
    /// Unseal error
    #[error(transparent)]
    UnsealError(#[from] DecryptionError),
    /// An error ocurred enqueueing in a fan out queue
    #[error(transparent)]
    EnqueueError(#[from] EnqueueError),
    /// An error ocurred while sending a message to the network
    #[error("An error ocurred while sending a message to the network")]
    NetworkError(N::NetworkError),
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid response
    #[error("Invalid response")]
    InvalidResponse,
}

/// Error enqueuing a fanned-out message.
#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum EnqueueError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError, // E.g. an error while encoding a message before enqueing it.
    /// Error in the underlying storage provider
    #[error("Error in the underlying storage provider")]
    Storage,
    /// Error sending push notification.
    #[error("Error sending push notification.")]
    PushNotificationError,
}

// === Internal errors ===

#[derive(Debug, Error)]
pub(super) enum GenerateAndStoreError {
    #[error("Error generating signature keypair")]
    KeyGenerationError(#[from] KeyGenerationError),
    #[error("Error storing key")]
    StorageError(#[from] StorageError),
}
