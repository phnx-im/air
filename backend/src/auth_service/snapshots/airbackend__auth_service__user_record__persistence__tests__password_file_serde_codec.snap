---
source: backend/src/auth_service/user_record.rs
expression: "AirCodec::to_vec(&*PASSWORD_FILE).unwrap()"
extension: cbor
snapshot_kind: binary
---
